<a href="https://colab.research.google.com/github/shanmughamkarthik-eng/python_practice/blob/main/loops_and_control_structures_assignments_6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Problem 6.8: Simple Calculator (Medium)

Create a simple calculator program that:

1. Shows a menu of operations: Add, Subtract, Multiply, Divide, Exit
2. Takes two numbers as input
3. Performs the selected operation
4. Shows the result
5. Repeats until the user chooses to exit
6. Handles division by zero error

Hint: Use while loop. Display menu, get choice. Use if-elif to perform operation. Check if divisor is 0 before division.

In [ ]:
class calculator:
  def __init__(self, number_a = 0, number_b = 0):
    self.number_a = number_a
    self.number_b = number_b

  def addition(self):
    return self.number_a + self.number_b

  def subtraction(self):
    return self.number_a - self.number_b

  def multiplication(self):
      return self.number_a * self.number_b

  def division(self):
      return self.number_a / self.number_b if self.number_b != 0 else "Error: Division by zero"



print("Simple Calculator")

calc = calculator()

while True:
  print("\n1. Add\n2. Subtract\n3. Multiply\n4. Divide\n5. Exit")
  choice = input("\nEnter your choice: ")

  if choice == "5":
    print("Exiting...")
    break

  elif choice in ["1", "2", "3", "4"]:
    try:
      calc.number_a = float(input("Enter first number: "))
      calc.number_b = float(input("Enter second number: "))

      if choice == "1":
        print("Result: ", calc.addition())

      elif choice == "2":
        print("Result: ", calc.subtraction())

      elif choice == "3":
        print("Result: ", calc.multiplication())

      elif choice == "4":
        print("Result: ", calc.division())

    except ValueError:
      print("Invalid input. Please enter a valid number.")

  else:
    print("Invalid choice. Please select a valid option.")


Simple Calculator

1. Add
2. Subtract
3. Multiply
4. Divide
5. Exit

Enter your choice: 1
Enter first number: 20
Enter second number: 212
Result:  232.0

1. Add
2. Subtract
3. Multiply
4. Divide
5. Exit

Enter your choice: 2
Enter first number: 23
Enter second number: 12
Result:  11.0

1. Add
2. Subtract
3. Multiply
4. Divide
5. Exit

Enter your choice: 3
Enter first number: 12
Enter second number: 13
Result:  156.0

1. Add
2. Subtract
3. Multiply
4. Divide
5. Exit

Enter your choice: 4
Enter first number: 123
Enter second number: 0
Result:  Error: Division by zero

1. Add
2. Subtract
3. Multiply
4. Divide
5. Exit

Enter your choice: 123
Invalid choice. Please select a valid option.

1. Add
2. Subtract
3. Multiply
4. Divide
5. Exit

Enter your choice: 4
Enter first number: 123
Enter second number: 2
Result:  61.5

1. Add
2. Subtract
3. Multiply
4. Divide
5. Exit

Enter your choice: 5
Exiting...


#Section 7: Bonus Challenges
##Problem 7.1: Collatz Conjecture (Medium)
The Collatz conjecture is one of the most famous unsolved problems in mathematics. The rules are:

1. Start with any positive integer n
2. If n is even, divide it by 2
3. If n is odd, multiply it by 3 and add 1
4. Repeat until you reach 1

###Write a program that:
*  Takes a number as input
*  Shows the complete sequence until it reaches 1
*  Counts how many steps it took

Example: Starting with 6: 6 → 3 → 10 → 5 → 16 → 8 → 4 → 2 → 1 (8 steps)

Hint: While n != 1: if n%2==0, n=n//2, else n=3*n+1. Count steps and print each value.

In [ ]:
print("Collatz Conjecture")

number = int(input("Enter a number: "))
steps = 0
print(f"\n{number}", end="")

while number!=1:
  if number % 2 == 0:
    number = number/2

  else:
    number = (3* number)+1

  print(f" -> {int(number)}", end="")
  steps += 1

print(f" ({steps} steps)")

Collatz Conjecture
Enter a number: 11

11 -> 34 -> 17 -> 52 -> 26 -> 13 -> 40 -> 20 -> 10 -> 5 -> 16 -> 8 -> 4 -> 2 -> 1 (14 steps)


###Problem 7.2: Caesar Cipher (Hard)
Write a program that implements a Caesar cipher to encrypt and decrypt messages.

The Caesar cipher shifts each letter by a fixed number of positions in the alphabet.

Example: With shift=3
*  'A' becomes 'D'
*  'Hello' becomes 'Khoor'

Your program should:

1. Ask for a message and shift value
2. Provide options to encrypt or decrypt
3. Handle both uppercase and lowercase letters
4. Keep non-alphabetic characters unchanged

Hint: Loop through each character. Use ord() to get ASCII value, shift it, use chr() to convert back. Use modulo 26 for wrapping. Check if char.isalpha().

In [ ]:
def get_shift_value():
  while True:
    try:
      return int(input("Enter the shift value: "))
    except ValueError:
      print("Invalid value. Please enter a shift value number.")

def caesar_cipher(message, shift_value):
  result = ""
  for char in message:
    if char.isupper():
        result += chr((ord(char) + shift_value - 65) % 26 + 65)

    elif char.islower():
        result += chr((ord(char) + shift_value - 97) % 26 + 97)

    else:
        result += char
  return result

# ------------------------

print("Caesar Cipher\n")

print("1. Encrypt")
print("2. Decrypt")
print("0. Quit")

while True:
  try:
    choice = int(input("\nEnter your choice: "))
  except ValueError:
    print("Invalid input. Please enter a valid number.")
    continue

  if choice == 0:
    print("Exiting")
    break

  elif choice == 1:
    message = input("Enter the message: ")
    shift_value = get_shift_value()
    print(f"Encrypted message: {caesar_cipher(message, shift_value)}")

  elif choice == 2:
    message = input("Enter the message: ")
    shift_value = get_shift_value()
    print(f"Decrypted message: {caesar_cipher(message, -shift_value)}")

  else:
    print("Invalid choice. Please select a valid option. eg., 0, 1 or 2")

Caesar Cipher

1. Encrypt
2. Decrypt
0. Quit

Enter your choice: w
Invalid input. Please enter a valid number.

Enter your choice: 9
Invalid choice. Please select a valid option. eg., 0, 1 or 2

Enter your choice: 1
Enter the message: HelloZ@125
Enter the shift value: 8
Encrypted message: PmttwH@125

Enter your choice: 2
Enter the message: PmttwH@125
Enter the shift value: 8
Decrypted message: HelloZ@125

Enter your choice: 1
Enter the message: hello
Enter the shift value: q
Invalid value. Please enter a shift value number.
Enter the shift value: 4
Encrypted message: lipps

Enter your choice: 2
Enter the message: lipps
Enter the shift value: d
Invalid value. Please enter a shift value number.
Enter the shift value: 4
Decrypted message: hello

Enter your choice: 0
Exiting


####Problem 7.3: ATM Machine Simulator (Hard)
#####Create an ATM machine simulator with the following features:

1. Initial balance: 10000
2. PIN: 1234 (3 attempts allowed)
3. Menu options:
  *   Check Balance
  *   Withdraw Money (check for sufficient balance and valid amount)
  *   Deposit Money
  *   Change PIN
  *   Exit
4. Keep transaction history (last 5 transactions)
5. Input validation for all operations

Hint: Use variables for balance, pin. Loop for PIN attempts. Use while loop for menu. Use list to store transaction history. Validate amounts > 0 and balance >= withdrawal.

In [ ]:
def withdraw_money():
  global balance
  global transaction_history

  while True:
    try:
      amount = float(input("Enter the amount: "))
    except ValueError:
      print("Please enter a numeric value.")
      continue

    if amount <= 0:
      print("Invalid amount. Please enter a positive number.")
      continue

    elif amount > balance:
      print("Insufficient balance.")
      continue

    else:
      balance -= amount
      transaction_history.append(f" - {amount}")
      print(f"Withdrawal successful. New balance: {balance}")
      return

#----------------------------------------------------------------

def deposit_money():
  global balance
  global transaction_history

  while True:
    try:
      amount = float(input("Enter the amount: "))
    except ValueError:
      print("Please enter a numeric value.")
      continue

    if amount <= 0:
      print("Invalid amount. Please enter a positive number.")
      continue

    else:
      balance += amount
      transaction_history.append(f" + {amount}")
      print(f"Deposit successful. New balance: {balance}")
      return

#-----------------------------------------------------------------

def check_balance():
  global balance
  global transaction_history

  print(f"Your balance is: {balance:,.2f}")

  if input("Do you want to see the last 5 transactions? (y/n): ").lower() == 'y':
     # This slices the list to get the last 5 items
    last_five = transaction_history[-5:]

    if not last_five:
      print("No transactions yet.")

    else:
      print("\n--- Recent Transactions ---")
      for tx in last_five:
        print(tx)
      return

  else:
    return

#-------------------------------------------------------------------

def change_pin():
  global pin
  global transaction_history
  attempts = 0

  while attempts < 3:
    try:
      current = int(input(f"Enter your current PIN (Attempt {attempts + 1}/3):"))

      if current != pin:
          print("Incorrect PIN. Authorization failed.")
          attempts += 1
          continue

      else:
        try:
          new_pin = int(input("Enter the new PIN: "))
          confirmation = int(input("Confirm the new PIN: "))

          if new_pin == confirmation:
            pin = new_pin
            print("PIN changed successfully.")
            return True

          else:
            print("Pin not matched")
            return False

        except ValueError:
          print("Please enter a numeric value.")
          continue

    except ValueError:
      print("Please enter a numeric value.")
      attempts += 1
      continue

  if attempts >= 3:
    print("Maximum number of attempts reached. Exiting.")
    return False

#-------------------------------------------------------------------------
def login():
  global pin
  attempts = 0

  while attempts < 3:
    try:
      entered_pin = int(input(f"Enter your PIN  (Attempt {attempts + 1}/3): "))
      if entered_pin == pin:
        print("Login successful.")
        return True

      else:
          print("Incorrect PIN. Authorization failed.")
          attempts += 1
          continue

    except ValueError:
      print("Please enter a numeric value.")
      attempts += 1
      continue

  return False

#-----------------------------------------------

print("ATM Machine Simulator")
balance = 10000
pin = 1234
transaction_history = []

if not login():
  print("Exiting")

else:
  while True:
    print("\n1. Check Balance")
    print("2. Withdraw Money")
    print("3. Deposit Money")
    print("4. Change PIN")
    print("5. Exit")

    try:
      choice = int(input("\nEnter your choice: "))
    except ValueError:
      print("Invalid input. Please enter a valid number.")
      continue

    if choice == 5:
      print("Exiting")
      break

    elif choice == 1:
      check_balance()

    elif choice == 2:
      withdraw_money()

    elif choice == 3:
      deposit_money()

    elif choice == 4:
      if not change_pin():
        break

    else: print("Invalid choice. Please select a valid option. eg., 1 to 5")

ATM Machine Simulator
Enter your PIN  (Attempt 1/3): 1232
Incorrect PIN. Authorization failed.
Enter your PIN  (Attempt 2/3): as
Please enter a numeric value.
Enter your PIN  (Attempt 3/3): 1234
Login successful.

1. Check Balance
2. Withdraw Money
3. Deposit Money
4. Change PIN
5. Exit

Enter your choice: 1
Your balance is: 10,000.00
Do you want to see the last 5 transactions? (y/n): n

1. Check Balance
2. Withdraw Money
3. Deposit Money
4. Change PIN
5. Exit

Enter your choice: 6
Invalid choice. Please select a valid option. eg., 1 to 5

1. Check Balance
2. Withdraw Money
3. Deposit Money
4. Change PIN
5. Exit

Enter your choice: 2
Enter the amount: qw
Please enter a numeric value.
Enter the amount: 1200
Withdrawal successful. New balance: 8800.0

1. Check Balance
2. Withdraw Money
3. Deposit Money
4. Change PIN
5. Exit

Enter your choice: 3
Enter the amount: qw
Please enter a numeric value.
Enter the amount: 3500
Deposit successful. New balance: 12300.0

1. Check Balance
2. Withdraw

    Enter your choice: 4
    Enter your current PIN (Attempt 1/3):1234
    Enter the new PIN: 1221
    Confirm the new PIN: 1222
    Pin not matched

    ATM Machine Simulator
    Enter your PIN  (Attempt 1/3): 1232
    Incorrect PIN. Authorization failed.
    Enter your PIN  (Attempt 2/3): 1221
    Incorrect PIN. Authorization failed.
    Enter your PIN  (Attempt 3/3): 1112
    Incorrect PIN. Authorization failed.
    Exiting

#Section 8: Additional If-Else Practice (30 Questions)
##Problem 8.1: Positive, Negative, or Zero (Easy)
Write a program that checks if a number is positive, negative, or zero.

Hint: Use if-elif-else: if n > 0 (positive), elif n < 0 (negative), else (zero).

In [ ]:
print("Positive, Negative, or Zero")

my_number = int(input("Enter a number: "))

if my_number > 0:
  print(f"{my_number} is a positive number")

elif my_number < 0:
    print(f"{my_number} is a negative number")

else:
  print(f"{my_number} is zero")

Positive, Negative, or Zero
Enter a number: -2
-2 is a negative number


    Positive, Negative, or Zero
    Enter a number: 0
    0 is zero

    Positive, Negative, or Zero
    Enter a number: 9
    9 is a positive number

##Problem 8.2: Even or Odd (Easy)
Write a program that determines if a number is even or odd.

Hint: If n % 2 == 0, it's even; otherwise odd.

In [ ]:
print("Even or Odd")

my_number = int(input("Enter a number: "))

if my_number % 2 == 0:
  print(f"{my_number} is an even number")

else:
    print(f"{my_number} is an odd number")

Even or Odd
Enter a number: 45
45 is an odd number


##Problem 8.3: Largest of Three Numbers (Easy)
Write a program to find the largest of three numbers.

Hint: Compare: if a >= b and a >= c, largest is a. Use nested if-else or use max() function.

In [ ]:
print("Largest of Three Numbers")

a = int(input("Enter first number: "))
b = int(input("Enter second number: "))
c = int(input("Enter third number: "))

if a >= b and a >= c:
    largest = a

elif b >= a and b >= c:
  largest = b

else:
  largest = c

print(f"The largest number is : {largest}")

Largest of Three Numbers
Enter first number: 34
Enter second number: 43
Enter third number: 12
The largest number is : 43


##Problem 8.4: Voting Eligibility (Easy)
Check if a person is eligible to vote (age >= 18).

Hint: If age >= 18, print eligible; else not eligible.

In [ ]:
print("Voting Eligibility Program")

age = int(input("Enter your age: "))

if age >= 18:
  print("You are eligible to vote")

else:
    print("You are not eligible to vote")

Voting Eligibility Program
Enter your age: 21
You are eligible to vote


##Problem 8.5: Character Type Checker (Medium)
Write a program that checks if a character is a vowel, consonant, digit, or special character.

Hint: Check if ch in 'aeiouAEIOU' (vowel), elif ch.isalpha() (consonant), elif ch.isdigit() (digit), else special.

In [ ]:
print("Character Type Checker")

character = input("\nEnter a character: ")

if character in "aeiouAEIOU":
    print(f"\nCharacter '{character}' is a vowel")

elif character.isalpha():
    print(f"\nCharacter '{character}' is a consonant")

elif character.isdigit():
    print(f"\nCharacter '{character}' is a digit")

else:
    print(f"\nCharacter '{character}' is a special character")


Character Type Checker

Enter a character: #

Character '#' is a special character


Character Type Checker

Enter a character: 12

Character '12' is a digit

Character Type Checker

Enter a character: f

Character 'f' is a consonant

Character Type Checker

Enter a character: u

Character 'u' is a vowel

###Problem 8.6: Discount Calculator (Medium)
Calculate discount based on purchase amount:
*   Below 1000: No discount
*   1000-5000: 10% discount
*   5000-10000: 20% discount
*   Above 10000: 30% discount


Hint: Use if-elif-else with ranges.

Calculate: final_amount = amount - (amount * discount_percentage / 100).

In [ ]:
print("Dicount Calculator")

try:
    purchase_amount = (float(input("Enter the purchase amount: ")))

    if purchase_amount < 0:
        print("Invalid input. Please enter a positive number.")

    else:
        if purchase_amount < 1000:
            discount_percentage = 0

        elif purchase_amount <= 5000:
            discount_percentage = 10

        elif purchase_amount <= 10000:
            discount_percentage = 20

        else:
            discount_percentage = 30

        final_amount = purchase_amount - (purchase_amount * discount_percentage / 100)

        print(f"Final amount after discount: ${final_amount:.2f}")

except ValueError:
    print("Invalid input. Please enter a valid number.")



Dicount Calculator
Enter the purchase amount: 30000
Final amount after discount: $21000.00


###Problem 8.7: Quadrant Finder (Medium)
Given x and y coordinates, determine which quadrant the point lies in (or on an axis).

Hint:   

    If x>0 and y>0: Quadrant I.
    If x<0 and y>0: Quadrant II.
    Similar for III and IV.
    Check for axes when x==0 or y==0.

In [ ]:
print("Quadrant Finder")

try:
    x = float(input("Enter the x-coordinate: "))
    y = float(input("Enter the y-coordinate: "))

    if x == 0 and y == 0:
        print("The point is at the origin.")

    elif x > 0 and y > 0:
        print("The point is in Quadrant I.")

    elif x < 0 and y > 0:
        print("The point is in Quadrant II.")

    elif x < 0 and y < 0:
        print("The point is in Quadrant III.")

    elif x > 0 and y < 0:
        print("The point is in Quadrant IV.")

    elif x == 0:
        print("The point is on the y-axis.")

    else:
        print("The point is on the x-axis.")


except ValueError:
    print("Invalid input. Please enter valid numbers.")

Quadrant Finder
Enter the x-coordinate: 23
Enter the y-coordinate: 45
The point is in Quadrant I.


    Quadrant Finder
    Enter the x-coordinate: 0
    Enter the y-coordinate: 1
    The point is on the y-axis.

    Quadrant Finder
    Enter the x-coordinate: 0
    Enter the y-coordinate: 0
    The point is at the origin.

###Problem 8.8: Age Category (Easy)
Categorize age as: Child (0-12), Teenager (13-19), Adult (20-59), Senior (60+).

Hint: Use if-elif-else with age ranges. Check from Child to Senior in order.

In [ ]:
print("Age Category")

try:
    age = int(input("Enter the age: "))

    if age < 0:
        print("please enter a valid age")

    elif age <= 12:
        print("Child")

    elif age <= 19:
        print("Teenager")

    elif age <= 59:
        print("Adult")

    else:
        print("Senior")

except ValueError:
    print("Invalid input. Please enter valid numbers.")

Age Category
Enter the age: 62
Senior


    Age Category
    Enter the age: ad
    Invalid input. Please enter valid numbers.

    Age Category
    Enter the age: -1
    please enter a valid age

###Problem 8.9: Temperature Converter (Medium)
Convert temperature between Celsius and Fahrenheit based on user choice.

#####C to F: F = (C × 9/5) + 32
#####F to C: C = (F - 32) × 5/9
Hint: Ask user for choice (1 or 2). Use if-else to apply appropriate formula. Use parentheses correctly in calculations.

In [ ]:
print("Temperature Converter")
try:
    temperature = float(input("Enter the temperature: "))
    choice = int(input("Enter 1 for Celsius to Fahrenheit or 2 for Fahrenheit to Celsius: "))

    if choice == 1:
        fahrenheit = (temperature * 9/5) + 32
        print(f"{temperature} C is equal to {fahrenheit:.2f} F")

    elif choice == 2:
        celsius = (temperature - 32) * 5/9
        print(f"{temperature} F is equal to {celsius:.2f} C")

    else:
        print("Invalid choice. Please enter 1 or 2.")

except ValueError:
    print("Invalid input. Please enter a valid number")

Temperature Converter
Enter the temperature: 45
Enter 1 for Celsius to Fahrenheit or 2 for Fahrenheit to Celsius: 2
45.0 F is equal to 7.22 C


####Problem 8.10: Password Strength Checker (Medium)
Check password strength:
*   Weak: length < 6
*   Medium: length 6-10
*   Strong: length > 10

Hint: Use len(password) to get length. Compare with if-elif-else.

In [ ]:
print("Password Strength Checker")

password = input("Enter the password: ")

if len(password) < 6:
    print("password is Weak")

elif len(password) <= 10:
    print("password is Medium")

else:
    print("password is Strong")

Password Strength Checker
Enter the password: waswerf
password is Medium


####Problem 8.11: Divisibility Checker (Easy)
Check if a number is divisible by both 3 and 5, only 3, only 5, or neither.

Hint: Check if n%3==0 and n%5==0 (both), elif n%3==0 (only 3), elif n%5==0 (only 5), else (neither).

In [ ]:
print("Divisibility checker")
try:
    my_number = int(input("Enter a number: "))

    if my_number == 0:
        print("Invalid input. Please enter a non-zero number")

    elif my_number % 3 == 0 and my_number % 5 == 0:
        print(f"{my_number} is divisible by both 3 and 5")

    elif my_number % 3 == 0:
        print(f"{my_number} is divisible by 3")

    elif my_number % 5 == 0:
        print(f"{my_number} is divisible by 5")

    else:
        print(f"{my_number} is not divisible by 3 or 5")

except ValueError:
    print("Invalid input. Please enter a valid number")

Divisibility checker
Enter a number: 15
15 is divisible by both 3 and 5


####Problem 8.12: Movie Ticket Pricing (Medium)
Calculate ticket price:
*   Children (< 12): $10

*   Adults (12-59): $15

*    Seniors (60+): $12

*   Weekend: Add $3 to all prices

Hint: First determine base price by age. Then check if weekend and add $3 if true.

In [ ]:
print("Movie Ticket Pricing")

try:
    user_age = int(input("Enter your age: "))
    is_weekend = input("Is it a weekend? (yes/no): ").lower()

    if user_age < 0:
        print("Invalid input. Please enter a valid age")

    else:
        if user_age < 12:
            base_price = 10

        elif user_age <= 59:
            base_price = 15

        else:
            base_price = 12

        if is_weekend == "yes":
            total_price = base_price + 3
            print(f"Total price: ${total_price:.2f}")

        else:
            print(f"Total price: ${base_price:.2f}")

except ValueError:
    print("Invalid input. Please enter a valid number")


Movie Ticket Pricing
Enter your age: 35
Is it a weekend? (yes/no): yes
Total price: $18.00


####Problem 8.13: Letter Grade to GPA (Medium)
Convert letter grade to GPA: A=4.0, B=3.0, C=2.0, D=1.0, F=0.0

Hint: Use if-elif-else or dictionary: grade_dict = {'A': 4.0, 'B': 3.0, ...}. Then print grade_dict[letter].

In [ ]:
print("Letter Grade to GPA")
grade_dict = {'A':4.0, 'B':3.0, 'C':2.0, 'D':1.0, 'F':0.0}

try:
    letter_grade = input("Enter the letter grade: ").upper()

    if letter_grade == "A":
        print(f"GPA: {grade_dict[letter_grade]}")

    elif letter_grade == "B":
        print(f"GPA: {grade_dict[letter_grade]}")

    elif letter_grade == "C":
        print(f"GPA: {grade_dict[letter_grade]}")

    elif letter_grade == "D":
        print(f"GPA: {grade_dict[letter_grade]}")

    elif letter_grade == "F":
        print(f"GPA: {grade_dict[letter_grade]}")

    else:
        print("Invalid input. Please enter a valid letter grade")

except ValueError:
    print("Invalid input. Please enter a valid letter grade")

Letter Grade to GPA
Enter the letter grade: a
GPA: 4.0


####Problem 8.14: Electricity Bill Calculator (Medium)
Calculate bill based on units consumed:

*   0-100: $0.50 per unit

*   101-200: $0.75 per unit

*   201-300: $1.20 per unit

*   Above 300: $1.50 per unit


Hint: Use if-elif-else to determine rate based on units.

Calculate: bill = units * rate.

In [ ]:
print("Electricity Bill Calculator")
try:
    units_consumed = int(input("Enter the number of units consumed: "))

    if units_consumed < 0:
        print("Invalid input. Please enter a non-negative number")

    else:
        if units_consumed <100:
            rate = 0.50

        elif units_consumed <= 200:
            rate = 0.75

        elif units_consumed <= 300:
            rate = 1.20

        else:
            rate = 1.50

        bill = units_consumed * rate
        print(f"Electricity bill: ${bill:.2f}")

except ValueError:
    print("Invalid input. Please enter a valid number")

Electricity Bill Calculator
Enter the number of units consumed: 100
Electricity bill: $75.00


####Problem 8.15: Traffic Light System (Easy)
Based on light color, print action: Red=Stop, Yellow=Caution, Green=Go

Hint: Use if-elif-else or dictionary to map color to action.

In [ ]:
import random
import time

def get_random_color(last_color):
    colors = ["Red", "Yellow", "Green"]

    if last_color in colors:
        colors.remove(last_color)
    return random.choice(colors)


print("Traffic Light system")
last_light = None

try:
    while True:
        light_color = get_random_color(last_light)
        last_light = light_color

        if light_color == "Red":
            print(f"{light_color} signal. Stop")
            time.sleep(3)

        elif light_color == "Yellow":
            print(f"{light_color} signal. Caution")
            time.sleep(1)

        elif light_color == "Green":
            print(f"{light_color} signal. Go")
            time.sleep(3)

except KeyboardInterrupt:
    print("Program terminated by user")

Traffic Light system
Red signal. Stop
Green signal. Go
Red signal. Stop
Yellow signal. Caution
Red signal. Stop
Yellow signal. Caution
Red signal. Stop
Yellow signal. Caution
Red signal. Stop
Yellow signal. Caution
Green signal. Go
Red signal. Stop
Green signal. Go
Yellow signal. Caution
Red signal. Stop
Yellow signal. Caution
Green signal. Go
Red signal. Stop
Program terminated by user


####Problem 8.16: Smallest of Three Numbers (Easy)
Find the smallest of three numbers.

Hint: Similar to Problem 8.3, but check for smallest. Or use min(a, b, c).

In [ ]:
print("smallest of Three Numbers")
try:
    a = int(input("Enter first number: "))
    b = int(input("Enter second number: "))
    c = int(input("Enter third number: "))

    smallest = min(a, b, c)

    print(f"The smallest number is : {smallest}")

except ValueError:
    print("Invalid input. Please enter a valid number")

smallest of Three Numbers
Enter first number: 10
Enter second number: 23
Enter third number: 3
The smallest number is : 3


####Problem 8.17: Number in Range (Easy)
Check if a number is between 1 and 100 (inclusive).

Hint: Check if 1 <= n <= 100.

In [ ]:
print("Number in Range")
try:
    number = int(input("Enter a number: "))

    if 1 <= number <= 100:
        print(f"{number} is in the range of 1 to 100")

    else:
        print(f"{number} is not in the range of 1 to 100")

except ValueError:
    print("Invalid input. Please enter a valid number")

Number in Range
Enter a number: 10
10 is in the range of 1 to 100


####Problem 8.18: Absolute Value (Easy)
Calculate the absolute value of a number without using built-in functions.

Hint: If n < 0, return -n; else return n.

In [ ]:
print("Absolute Value")
try:
    number = float(input("Enter a number: "))

    if number < 0:
        absolute_value = -number

    else:
        absolute_value = number

    print(f"The absolute value of '{number}' is:  {absolute_value}")

except ValueError:
    print("Invalid input. Please enter a valid number")

Absolute Value
Enter a number: -12
The absolute value of '-12.0' is:  12.0


####Problem 8.19: Days in Month (Medium)
Given month number (1-12), print the number of days. Consider leap year for February.

Hint: Use if-elif for each month. For Feb (month==2), check if leap year (use Problem 1.2 logic) to return 29 or 28.

In [ ]:
def is_leap_year(year):
    if (year % 4 == 0 and year % 100 != 0) or (year % 400 == 0):
        return True
    else:
        return False


print("Days in Month")

try:
    month = int(input("Enter the month number (1-12): "))
    year = int(input("Enter the year: "))

    if month < 1 or month > 12:
        print("Invalid input. Please enter a valid month number (1-12)")

    else:
        if month in [1, 3, 5, 7, 8, 10, 12]:
            print(f"Month {month} has 31 days")

        elif month in [4, 6, 9, 11]:
            print(f"Month {month} has 30 days")

        else:
            if is_leap_year(year):
                print(f"year {year} is a leap year. Month {month} has 29 days")

            else:
                print(f"Month {month} has 28 days")

except ValueError:
    print("Invalid input. Please enter a valid number")



Days in Month
Enter the month number (1-12): 2
Enter the year: 2004
year 2004 is a leap year. Month 2 has 29 days


####Problem 8.20: Profit or Loss Calculator (Easy)
Calculate profit or loss given cost price and selling price.

Hint: If selling_price > cost_price: profit = selling_price - cost_price. Else if less: loss. Else: no profit no loss.

In [ ]:
print("Profit or Loss Calculator")
try:
    selling_price = float(input("Enter the selling price: "))
    cost_price = float(input("Enter the cost price: "))

    if selling_price > cost_price:
        profit = selling_price - cost_price
        print(f"Profit: Rs.{profit:.2f}")

    elif selling_price < cost_price:
        loss = cost_price - selling_price
        print(f"Loss: Rs.{loss:.2f}")

    else:
        print("No profit no loss")

except ValueError:
    print("Invalid input. Please enter a valid number")

Profit or Loss Calculator
Enter the selling price: 3000
Enter the cost price: 2400
Profit: Rs.600.00


####Problem 8.21: Uppercase or Lowercase (Easy)
Check if a character is uppercase or lowercase.

Hint: Use ch.isupper() or ch.islower(), or compare ASCII values.

In [ ]:
print("Uppercase or Lowercase")
try:
    character = input("Enter a character: ")

    if len(character) != 1:
        print("Invalid input. Please enter a single character")

    else:
        if character.isupper():
            print(f"'{character}' is an uppercase letter")

        elif character.islower():
            print(f"'{character}' is a lowercase letter")

        else:
            print(f"'{character}' is not an uppercase or lowercase letter")
            print("Please enter an uppercase or lowercase letter")

except ValueError:
    print("Invalid input. Please enter a valid character")

Uppercase or Lowercase
Enter a character: A
'A' is an uppercase letter


####Problem 8.22: Roots of Quadratic Equation (Hard)
Find the nature of roots of quadratic equation ax² + bx + c = 0:

1. Real and distinct (discriminant > 0)
2. Real and equal (discriminant = 0)
3. Imaginary (discriminant < 0)

Discriminant = b² - 4ac

Hint: Calculate discriminant = b**2 - 4ac. Use if-elif-else to check if > 0, == 0, or < 0.

In [ ]:
def discriminant(a, b, c):
    return b**2 - 4*a*c

print("Roots of Quadratic Equation")
try:
    a = int(input("Enter the value of a: "))
    b = int(input("Enter the value of b: "))
    c = int(input("Enter the value of c: "))

    if a == 0:
        print("Invalid input. 'a' cannot be zero")

    else:
        d = discriminant(a, b, c)

        if d > 0:
            print("Real and distinct roots")

        elif d == 0:
            print("Real and equal roots")

        else:
            print("Imaginary roots (complex roots)")

except ValueError:
    print("Invalid input. Please enter a valid number")

Roots of Quadratic Equation
Enter the value of a: 1
Enter the value of b: 3
Enter the value of c: 6
Imaginary roots (complex roots)


####Problem 8.23: Bonus Calculation (Medium)
Calculate employee bonus:
1. Salary < 10000: 10% bonus
2. Salary 10000-20000: 15% bonus
3. Salary > 20000: 20% bonus
4. If years of service > 5: Add extra 5%

Hint: Determine bonus percentage based on salary.

If years > 5, add 5 to bonus percentage.

Calculate: bonus = salary * bonus_percentage / 100.

In [ ]:
print("Bonus Calculation")
try:
    salary = float(input("Enter the salary: "))
    years_of_service = int(input("Enter the years of service: "))

    if salary < 0:
        print("Invalid input. Please enter a non-negative salary")

    else:
        if salary < 10000:
            bonus_percentage = 10

        elif salary <= 20000:
            bonus_percentage = 15

        else:
            bonus_percentage = 20

        if years_of_service <= 0:
            print("Invalid input. Please enter a positive number of years of service")

        else:
            if years_of_service > 5:
                bonus_percentage += 5

            bonus = salary * bonus_percentage / 100
            print(f"Total Bonus Percentage: {bonus_percentage}%")
            print(f"Bonus Amount: Rs.{bonus:.2f}")
            print(f"Total Pay: Rs.{(salary + bonus):.2f}")

except ValueError:
    print("Invalid input. Please enter a valid number")

Bonus Calculation
Enter the salary: 10000
Enter the years of service: 6
Total Bonus Percentage: 20%
Bonus Amount: Rs.2000.00
Total Pay: Rs.12000.00


####Problem 8.24: Water State (Easy)
Determine water state based on temperature:
1. Below 0°C: Ice
2. 0-100°C: Water
3. Above 100°C: Steam

Hint: If temp < 0: Ice, elif temp <= 100: Water, else: Steam.

In [ ]:
print("Water State")
try:
    temperature = float(input("Enter the temperature in degrees Celsius: "))

    if temperature < 0:
        print("Water state is : Ice")

    elif temperature <= 100:
        print("Water state is : Water")

    else:
        print("Water state is : Steam")

except ValueError:
    print("Invalid input. Please enter a valid number")

Water State
Enter the temperature in degrees Celsius: 101
Water state is : Steam


####Problem 8.25: Alphabet or Digit (Easy)
Check if a character is an alphabet or a digit.

Hint: Use ch.isalpha() to check alphabet, ch.isdigit() for digit.

In [ ]:
print("Alphabet or Digit")

character = input("Enter a character: ")
try:
    if len(character) != 1:
        print("Invalid input. Please enter a single character")

    else:
        if character.isalpha():
            print(f"'{character}' is an alphabet")

        elif character.isdigit():
            print(f"'{character}' is a digit")

        else:
            print(f"'{character}' is not an alphabet or a digit")
            print("Please enter an alphabet or a digit")

except ValueError:
    print("Invalid input. Please enter a valid")

Alphabet or Digit
Enter a character: A
'A' is an alphabet


Problem 8.26: Tax Calculator (Medium)
Calculate income tax:

1. Income 0-250000: No tax
2. Income 250001-500000: 5% tax
3. Income 500001-1000000: 10% tax
4. Income above 1000000: 20% tax

Hint: Use if-elif-else to determine tax rate. Calculate: tax = income * tax_rate / 100.

In [6]:
print("Tax Calculator")
try:
    income = float(input("Enter the income: "))

except ValueError:
    print("Invalid input. Please enter a valid number")

if income < 0:
    print("Invalid input. Please enter a non-negative income")

else:
    if income <= 250000:
        tax_rate = 0

    elif income <= 500000:
            tax_rate = 5

    elif income <= 1000000:
            tax_rate = 10

    else:
            tax_rate = 20

    tax = income * tax_rate / 100

    print(f"Income Tax for {income} is : Rs.{tax:.2f}")


Tax Calculator
Enter the income: 6000000
Income Tax for 6000000.0 is : Rs.1200000.00
